Make basic community membership guidelines clearer (#3827)
kubeflow/website: content/en/docs/about/membership.md

+++
title = "Community Membership"
description = "Guidelines for contributing to Kubeflow"
weight = 30
aliases = ["/docs/membership/"]
+++

> **Note:** This document is a work in progress

Reviewer: @anishasthana Do you want to make this change for your Notes ?

> **Note:** This document is a work in progress

Author: Done, I think!

This document outlines the various responsibilities of contributor roles in Kubeflow. Kubeflow is divided into working groups that have stewardship over different subprojects/repositories

Responsibilities for most roles are scoped to these repositories.

> | Role | Responsibilities | Requirements | Defined by |

Reviewer 2: Can we also add the steering committee here?

Author: Is there documentation anywhere for the KSC that I can pull from?

Reviewer 2: The only doc is available here https://github.com/kubeflow/community/blob/master/KUBEFLOW-STEERING-COMMITTEE.md

Author: I'm concerned about having conflicting documents.. Should I just link to the KSC document from here?

Reviewer 2: you can provide a short description of the role under "responsibilities" and link back to KSC doc for "requirement" and "defined by"

| -----| ---------------- | ------------ | -------|
| Member | Active contributor in the community | Sponsored by 2 reviewers and multiple contributions to the project | Kubeflow GitHub org member|

> | Reviewer | Review contributions from other members | History of review and authorship in a repository | [OWNERS] file reviewer entry |

Author: OWNERS files for each repository would be different right? Does it make sense to link the one for this repository?

> | Reviewer | Review contributions from other members | History of review and authorship in a repository | [OWNERS] file reviewer entry |

Reviewer: I don't think we should link the actual OWNERS files, this is just reference to the explanation: What is OWNERS file?

| Approver | Contributions acceptance approval| Highly experienced active reviewer and contributor to a repository | [OWNERS] file approver entry|

> | WG Lead  | Provides technical leadership for a Working Group | Have sufficient domain knowledge to provide effective technical leadership | [wgs.yaml] entry |

Reviewer 2: It is already linked here.

> | WG Lead  | Provides technical leadership for a Working Group | Have sufficient domain knowledge to provide effective technical leadership | [wgs.yaml] entry |

Reviewer: The link doesn't work for me.
![Screenshot 2024-09-30 at 16 20 50](https://github.com/user-attachments/assets/b08dc921-f466-4da2-a028-1241dec0926f)

> | WG Lead  | Provides technical leadership for a Working Group | Have sufficient domain knowledge to provide effective technical leadership | [wgs.yaml] entry |

Author: ![image](https://github.com/user-attachments/assets/a116cbfc-6b4b-4642-a0b6-d466f189e04e)
Fixed. Good catch @andreyvelich

| WG Chair | Provides overall leadership for a Working Group | Have sufficient domain knowledge to provide effective leadership | [wgs.yaml] entry |

**Note:** Detail documentation for Working Group structure and responsibilities can be found at [wg-governance.md](https://github.com/kubeflow/community/blob/master/wgs/wg-governance.md)

## New contributors

> [New contributors] should be welcomed to the community by existing members, helped with PR workflow, and directed to relevant documentation and communication channels.

Reviewer: ```suggestion
[New contributors] should be welcomed to the community by existing members, helped with PR workflow, and directed to the relevant documentation and communication channels.
```

## Established community members

Established community members are expected to demonstrate their adherence to the principles in this document, familiarity with project organization, roles, policies, procedures, conventions, etc., and technical and/or writing ability. Role-specific expectations, responsibilities, and requirements are enumerated below.

## Member

Members are *[continuously active]* contributors in the community. They can have issues and PRs assigned to them and tests are automatically run for their PRs. Members are expected to remain active contributors to the community.

**Defined by:** Member of the Kubeflow GitHub organization

### Requirements

- Enabled two-factor authentication on their GitHub account

> - Have made **at least** 2-3 [contributions] to the project or community.

Reviewer 3: ```suggestion
- Have made **at least** 2-3 [contributions] to the project or community, preferably related to more than a single working group.
```

Author: I wonder if this is unnecessarily "high" bar to entry. Some prospective members could be interested in specific subsets of the Kubeflow project.

Reviewer 4: +1. I think for some areas people may choose to just stay in one domain (e.g., training). Probably not necessary to require this.

Reviewer 3: ack, and agree

> - Have made **at least** 2-3 [contributions] to the project or community.

Reviewer: Should we link more concrete docs here. For example:
Have made **at least** 2-3 [code contributions](https://contribute.cncf.io/contributors/getting-started/#code-contributors) or [non-code contributions](https://contribute.cncf.io/contributors/getting-started/#non-code-contributors).
WDYT ?

Author: Added!

> - Have read the [contributor guide]

Reviewer: ```suggestion
- Have read the [contributor guide.]
```

> - Sponsored by 2 reviewers. **Note the following requirements for sponsors**:

Reviewer 3: ```suggestion
- Sponsored by 2 reviewers, preferably from different working groups. **Note the following requirements for sponsors**:
```

Reviewer: I would relax this rule @tarilabs. Usually, it is very hard to understand all different Kubeflow Working Groups for new members. So I would just say Sponsored by 2 Kubeflow members here.
WDYT @tarilabs @terrytangyuan @anishasthana ?

Reviewer 3: works for me!

Reviewer 3: ```suggestion
- Sponsored by 2 Kubeflow members. **Note the following requirements for sponsors**:
```

reflecting https://github.com/kubeflow/website/pull/3827#discussion_r1777153763

Author: I'm good with Kubeflow members too!

Reviewer 3: awesome

> - **[Open an issue][membership request] against the kubeflow/internal-acls repo**

Reviewer: This link doesn't work

  - Ensure your sponsors are @mentioned on the issue
- **Open a pull request against the kubeflow/internal-acls repo**
  - Complete every item on the checklist ([preview the current version of the template][membership template])
  - Make sure that the list of contributions included is representative of your work on the project.
- Have your sponsoring reviewers reply confirmation of sponsorship
- Once your sponsors have responded, your request will be reviewed by the Kubeflow team. Any missing information will be requested

### Responsibilities

> - Responsive to issues and PRs assigned to them

Reviewer: Maybe we should also say, being active in the Kubeflow community, joining working groups meetings, Slack discussions, and project discussions.

> - Responsive to mentions of teams they are members of

Reviewer: This is not necessary since members can be without GitHub team.

Author: On the flipside, if a member _is_ a part of a team they should be responsive right? I'd rather be explicit

Reviewer: Sure, since you change this sentence, it looks good
```
Responsive to mentions of any teams they may be members of

- Active owner of code they have contributed (unless ownership is explicitly transferred)
  - Code is well tested
  - Tests consistently pass
  - Addresses bugs or issues discovered after code is accepted

> - Subscribed to <kubeflow-discuss@googlegroups.com>

Reviewer: Let's use this links to join: https://groups.google.com/g/kubeflow-discuss

**Note:** members who frequently contribute code are expected to proactively perform code reviews and work towards becoming a primary *reviewer* for the subproject that they are active in.

> ### Privileges

Reviewer: Should we also say, ability to be a Kubeflow release manager?
WDYT @kubeflow/release-team ?

Author: I've added it to the privileges, but happy to remove it if the release team feels otherwise.

- Members can do `/lgtm` on open PRs.
- They can be assigned to issues and PRs, and people can ask members for reviews with a `/cc @username`.
- Tests can be run against their PRs automatically. No `/ok-to-test` needed.

> - Members can do `/ok-to-test` for PRs that have a `needs-ok-to-test` label, and use commands like `/close` to close PRs as well.

Reviewer: I think, there are more commands than `/close`. Eventually, they can run any Prow command that members can run: https://prow.k8s.io/command-help

## Reviewer

Reviewers are able to review code for quality and correctness on some part of a subproject. They are knowledgeable about both the codebase and software engineering principles.

> **Defined by:** *reviewers* entry in an OWNERS file in a repo owned by the Kubeflow project.

Reviewer: ```suggestion
**Defined by:** *reviewers* entry in an OWNERS file in a repo owned by the Kubeflow organization.
```

Reviewer status can be scoped to either parts of the codebase or the root directory for the entire codebase

**Note:** Acceptance of code contributions requires at least one approver in addition to the assigned reviewers.

### Requirements

The following apply to the part of codebase for which one would be a reviewer in an [OWNERS] file.

- member for at least 3 months

> - Primary reviewer for at least 5 PRs to the codebase

Reviewer: What about answering user's questions in GitHub issues and Slack ?

Author: Added

> - Reviewed or merged at least 20 substantial PRs to the codebase

Reviewer: Should we have 15 here ? E.g. half of the PRs to become an approver.

- Knowledgeable about the codebase
- Sponsored by a subproject approver
  - With no objections from other approvers
  - Done through PR to update the OWNERS file
- May either self-nominate or be nominated by an approver in this subproject

> **Note:** Working Group Leads may nominate and approve `Reviewers` that don't meet these requirements due to exceptional circumstances. While acceptable in the short term, Working Group Leads should ensure that these `Reviewers` eventually meet the requirements

Reviewer: Should we numerate these exception ?

Author: I'm not actually aware of specific exceptions, this is more so in place if some of the existing projects have reviewers/owners that don't meet the bar _yet_. I'm sure there has usually been a good reason for a reviewer/approver to be appointed when not meeting the requirements.

I expect the need for exceptions to go down as the community grows

Reviewer: >I'm not actually aware of specific exceptions, this is more so in place if some of the existing projects have reviewers/owners that don't meet the bar yet. I'm sure there has usually been a good reason for a reviewer/approver to be appointed when not meeting the requirements.

Should we follow the rules in that case ? I feel that we should either cleanup those OWNERs files or enumerate the exceptions. WDYT @franciscojavierarceo @kubeflow/kubeflow-steering-committee ?

Author: I didn't want to disrupt the existing projects since these guidelines haven't explicitly been stated before. I'd rather we introduce a timeline (say, 3 months) for all projects to audit and clean up their OWNERS files. That being said, I'll defer to the KSC on this!

Reviewer 2: I don't have a preference here and don't think it's a blocker for this PR IMO.

Reviewer: I would suggest to remove this Note from the doc for now. We can always update it once we discuss those exceptions.
What do you think @terrytangyuan @anishasthana ?

> The following apply to the part of codebase for which one would be a reviewer in an [OWNERS] file (for repos using the bot).

Reviewer: What do you mean by for repos using the bot ?

Author: Ah, I was referring to Prow, but I can see how it's unclear. I will remove this.

> ### Responsibilities

Reviewer: Should we also say: Answer users questions in GitHub issues and Kubeflow Slack channels ?

Author: Added

- All responsiblities that community members have

> - Responsible for project quality control via [code reviews]

Reviewer: link doesn't work

  - Focus on code quality and correctness, including testing and factoring
  - May also review for more holistic issues, but not a requirement
- Expected to be responsive to review requests
- Assigned PRs to review related to subproject of expertise
- Assigned test bugs related to subproject of expertise

### Privileges

- All Privileges that community members have
- Code reviewer status may be a precondition to accepting large code contributions
- May get a badge on PR and issue comments

## Approver

Code approvers are able to both review and approve code contributions. While
code review is focused on code quality and correctness, approval is focused on
holistic acceptance of a contribution including: backwards / forwards
compatibility, adhering to API and flag conventions, subtle performance and

> correctness issues, interactions with other parts of the system, etc.

Reviewer: Maybe we should also say code test coverage.

> **Defined by:** *approvers* entry in an OWNERS file in a repo owned by the Kubeflow project.

Reviewer: ```suggestion
**Defined by:** *approvers* entry in an OWNERS file in a repo owned by the Kubeflow organization.
```

> Approver status is scoped to a part of the codebase.

Reviewer: Should this statement be the same as for reviewers ?

### Requirements

The following apply to the part of codebase for which one would be an approver in an [OWNERS] file (for repos using the bot).

> - Reviewer of the codebase for at least 3 months

Reviewer: What about 5 months ? Otherwise, we should relax the reviewer rule.

> - Reviewer of the codebase for at least 3 months

Reviewer: Should also approver be the reviewer before ?

- Primary reviewer for at least 10 substantial PRs to the codebase

> - Reviewed or merged at least 30 PRs to the codebase

Reviewer: I think, for approver we should add responsibility to merge some PRs, WDYT ?

Author: Added

Reviewer: @anishasthana Did you add that change ? I still can see this message:
```
Reviewed or merged at least 30 PRs to the codebase
```

> - Nominated by a WG Lead, Chair or owner

Reviewer: I would only leave WG Lead or Chair.

Author: I was thinking of situations where a root approver may nominate an approver for a subdirectory or similar. WG Leads and Chairs aren't necessarily involved with the day-to-day of every repository right?

Reviewer: In that case, we should explain what is root approver and sub-directory approver in terms of control of the projects.
As you can see, WG Leads should be involved into day-to-day actives for every repository: https://github.com/kubeflow/community/blob/master/wgs/wg-governance.md#all-leads 

That is why, I feel that for now we should delegate it to WG Leads or Chairs for simplicity.

Reviewer: Any thoughts @anishasthana @terrytangyuan ?

  - With no objections from other Leads or owners

>   - Done through PR to update the top-level OWNERS file

Reviewer: Should this be top-level OWNER ?

> **Note:** Working Group Leads may nominate and approve `Approvers` that don't meet these requirements due to exceptional circumstances. While acceptable in the short term, Working Group Leads should ensure that these `Approvers` eventually meet the requirements

Reviewer: Same question as for reviewer.

### Responsibilities

The following apply to the part of codebase for which one would be an approver
in an [OWNERS] file (for repos using the bot).

- All responsibilities that reviewers have
- Approver status may be a precondition to accepting large architectural contributions
- Demonstrate sound technical judgement
- Responsible for project quality control via code reviews
  - Focus on holistic acceptance of contribution such as dependencies with other features, backwards / forwards
    compatibility, API and flag definitions, etc
- Expected to be responsive to review requests
- Mentor contributors and reviewers

### Privileges

- All privileges that reviewers have
- May approve code contributions for acceptance

## Inactive members

*Members are continuously active contributors in the community.*

A core principle in maintaining a healthy community is encouraging active
participation. It is inevitable that people's focuses will change over time and
they are not expected to be actively contributing forever.

However, being a member of one of the Kubeflow GitHub organizations comes with
an elevated set of permissions. These capabilities should not be used by those

> that are not familiar with the current state of the Kubeflow project.

Reviewer: ```suggestion
that are not familiar with the current state of the Kubeflow organization.
```

Therefore members with an extended period (1 year) away from the project with no activity
will be removed from the Kubeflow GitHub Organizations and will be required to
go through the org membership process again after re-familiarizing themselves
with the current state.

### How inactivity is measured

Inactive members are defined as members of one of the Kubeflow Organizations

> with **no** contributions across any organization within 12 months.

Reviewer: Maybe we should explicitly say ""with no technical and non-technical contributions""

Reviewer 2: Isn't that the same as "with no contributions"?

Reviewer: I think, we can just call it contributions if we have separate section in our doc which explains: What is contribution ?

Reviewer 2: We can follow similar guideline to what Argo project is doing: https://github.com/argoproj/argoproj/blob/main/community/membership.md#how-inactivity-is-measured

Basically DevStats should give us a sense of activity.

Author: Added.

After an extended period away from the project with no activity
those members would need to re-familiarize themselves with the current state
before being able to contribute effectively.

## Credit

This set of guidelines is heavily inspired by the [Kubernetes membership guidelines](https://github.com/kubernetes/community/blob/master/community-membership.md?plain=1#community-membership).

[contributions]: https://contribute.cncf.io/contributors/getting-started/
[contributor guide]: https://www.kubeflow.org/docs/about/contributing/
[membership template]: https://github.com/kubeflow/internal-acls/blob/master/.github/ISSUE_TEMPLATE/join_org.md

> [New contributors]: ./CONTRIBUTING.md

Reviewer 5: This link is broken.

```suggestion
[New contributors]: /docs/about/contributing/
```

+++
title = "Community Membership"
description = "Guidelines for contributing to Kubeflow"
weight = 30
aliases = ["/docs/membership/"]
+++

{{< note >}}

> This document is a work in progress

Reviewer 4: do you want to remove this?

Author: The intent was to say 'this document is not final", but I'm good with removing it if this muddies the message
